src/grpc/dp_grpc_impl.c

```diff
@@ -502,6 +502,12 @@ static int dp_process_create_interface(struct dp_grpc_responder *responder)
 		goto route6_err;
 	}
 
+	if (!port->is_pf)
+		if (DP_FAILED(dp_port_meter_config(port, request->total_flow_rate_cap, request->public_flow_rate_cap))) {
```

Comment:
We should also store the total_flow and public_flow rates in the dp_port's dp_port_iface sub structure so the get and list GRPC calls can return what was set. This should eliminate the boolean flag in dp_port, as if public_flow is zero, it is not set and also the same for total_flow. The checks whether these are 0 or not can be also put behind a macro for better readibility.

And the second thing is if total_flow_rate_cap and public_flow_rate_cap are not set, we should not call dp_port_meter_config() at all. This will ensure the backward compatibility and the TAP device compatibility will be also ensured. (The caller should not set the metering parameters and our tests do not set it)

Reply:
thanks for the comments. I just summarise the the changes to address the above comments since they are highly related.
1) public_flow_rate and total_flow_rate  are added into dp_port_iface, and thus the get/list interface calls will also include flow rates information (corresponding Go part is also adapted);
2) soft_metering_enabled is removed as suggested due to the change in 1);
3) to handle TAP device in the test scenario, the approach to get current tap type is used inside dp_port_meter_config(), to be consistent with other part of the port config code handling test cases. 

Relying only the value of total_flow_rate_cap and public_flow_rate_cap can work. but I found that it is not sufficient in a common development case, and we still need to do a config on sysfs file.
 ```
./dpservice-cli init
./dpservice-cli create interface --id=vm-1 --vni=100 --ipv4=192.168.129.5 --ipv6=2002::123 --device=8a:00.0_representor_vf0 --total-meter-rate=1000
```
then we finish some tests and stop dpservice, then run it again. This time, we create the interface without any metering parameter.
```
./dpservice-cli create interface --id=vm-1 --vni=100 --ipv4=192.168.129.5 --ipv6=2002::123 --device=8a:00.0_representor_vf0
```
If we don't do a config on sysfs file and reset the value to 0, the configured total-meter-rate would still take effect, which is not nice.

Reply:
@byteocean 
Thanks. The changes for my last comments look good. I also agree with setting to zero, as the TAP device case is now also handled nicely in a different way.

There is still a problem though. You go to `err` label when the error `DP_GRPC_ERR_PORT_METER` occurs without doing any rollback so you left a "broken" interface state behind, if that error occurs. 

And the `dp_start_port `is also already called and it is not so easy to rollback the started port. My suggestion is to place the `dp_port_meter_config `before `dp_start_port `and arrange the error labels in such a way that `dp_start_port `failure can also rollback the `dp_port_meter_config `call.

Reply:
nice spot and thanks for the suggestion to change. done. I also looked the counter part in `dp_process_delete_interface`, nothing similar needs to be worried there.